torch/utils/data/datapipes/utils/common.py

```diff
@@ -20,6 +21,76 @@
 ]
 
 
+def validate_input_col(fn: Callable, input_col: Optional[Union[int, tuple, list]]):
+    """
+    Checks that function used in a map style datapipe works with the input column
+
+    Args:
+        fn: The function to check.
+        input_col: The input column to check.
+    Returns:
+        None.
+    Raises:
+        ValueError: If the function is not compatible with the input column.
+    """
+    sig = inspect.signature(fn)
+    if isinstance(input_col, (list, tuple)):
+        input_col_size = len(input_col)
+    else:
+        input_col_size = 1
+
+    fn_name = str(fn)
+
+    def get_non_default_params(signature: inspect.Signature):
+        excluded_types = [inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD]
+        return [p for p in signature.parameters.values() if p.default is p.empty and p.kind not in excluded_types]
+
+    def contains_var_positional(signature: inspect.Signature):
+        """Checks for the presences of *args in the signature"""
+        return any(p.kind == inspect.Parameter.VAR_POSITIONAL for p in signature.parameters.values())
+
+    def get_normal_params(signature: inspect.Signature):
+        """Finds any position or keyword arguments in the signature"""
+        return [p for p in signature.parameters.values() if p.kind is inspect.Parameter.POSITIONAL_OR_KEYWORD]
+
+    if len(sig.parameters) >= input_col_size:
+        non_default_params = get_non_default_params(sig)
+
+        if len(non_default_params) > input_col_size:
+
+            raise ValueError(
+                f"The function {fn_name} takes {len(non_default_params)} "
+                f"non-default parameters, but {input_col_size} are required."
+            )
+
+        non_default_kw_only = [p for p in non_default_params if p.kind == inspect.Parameter.KEYWORD_ONLY]
+
+        if len(non_default_kw_only) > 0:
+            raise ValueError(
+                f"The function {fn_name} takes {len(non_default_kw_only)} "
+                f"non-default keyword-only parameters, which is not allowed."
+            )
+
+        pos_or_kwonly = get_normal_params(sig)
+        non_default_pos_or_kwonly = list(filter(lambda p: p.default is p.empty, pos_or_kwonly))
```

Comment:
Noob question: What's the difference between `non_default_pos_or_kwonly` and `non_default_params`? Is this used by `partial` function?
Could we combine them together?

Reply:
This is an additional check for partial functions which have signature like `fn(d0, *, d1=1)`, and non-default-params does not pick this up correctly

Reply:
I tried removing this and it was not able to properly differentiate the partial functions from the normal cases. The overhead is not very big so maybe it can stay?

Reply:
Let me pull your PR and take a look.